pkg/generators/generator.go

```diff
@@ -161,47 +156,60 @@ func (g *Generator) start() {
 	}
 	grp.Go(func() error {
 		g.logger.Info("starting partition key generation loop")
-		routingKeyCreator := &routingkey.Creator{}
-		r := rand.New(rand.NewSource(g.seed))
-		var (
-			cntCreated uint64
-			cntEmitted uint64
-		)
+		g.routingKeyCreator = &routingkey.Creator{}
+		g.r = rand.New(rand.NewSource(g.seed))
 		for {
-			values := g.createPartitionKeyValues(r)
-			token, err := routingKeyCreator.GetHash(g.table, values)
-			if err != nil {
-				g.logger.Panic(errors.Wrap(err, "failed to get primary key hash").Error())
-			}
-			idx := token % g.partitionCount
-			partition := g.partitions[idx]
-			cntCreated++
+			g.fillAllPartitions()
 			select {
-			case partition.values <- &typedef.ValueWithToken{Token: token, Value: values}:
-				cntEmitted++
 			case <-gCtx.Done():
 				g.logger.Debug("stopping partition key generation loop",
-					zap.Uint64("keys_created", cntCreated),
-					zap.Uint64("keys_emitted", cntEmitted))
+					zap.Uint64("keys_created", g.cntCreated),
+					zap.Uint64("keys_emitted", g.cntEmitted))
 				return gCtx.Err()
-			default:
-				// This part is only get triggered when partition is full of values
-				// Which is signal to stop generating
-				// But if partitions values are not balanced, you can have case when one partition is full
-				// While other partitions are low on values
-				// To address this case before pausing generation we need to make sure that all partition are above the limit
-				if g.partitions.NeedMoreValues() {
-					g.waitForMoreValuesNeeded()
-				}
+			case <-g.wakeUpSignal:
 			}
 		}
 	})
 }
 
-func (g *Generator) createPartitionKeyValues(r *rand.Rand) []interface{} {
+// fillAllPartitions guarantees that each partition was tested to be full
+// at least once since the function started and before it ended.
+// In other words no partition will be starved.
+func (g *Generator) fillAllPartitions() {
+	pFilled := make([]bool, len(g.partitions))
+	allFilled := func() bool {
+		for _, filled := range pFilled {
+			if !filled {
+				return false
+			}
+		}
+		return true
+	}
```

Comment:
Let isolate this code into another function

Reply:
What's the rationale behind this?

Reply:
My initial thoughts was that it will allocate memory for the function, but i looked at the resulted ASM - optimization eats it all away. No worries there.